src/renderers/testing/ReactShallowRenderer.js

```diff
@@ -0,0 +1,121 @@
+/**
+ * Copyright 2013-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ * @providesModule ReactShallowRenderer
+ * @preventMunge
+ */
+
+'use strict';
+
+const React = require('react');
+const ReactTestRenderer = require('react-test-renderer');
+
+const emptyFunction = require('fbjs/lib/emptyFunction');
+const invariant = require('fbjs/lib/invariant');
+
+const ShallowNodeMockComponent = ({children}) =>
+  React.Children.toArray(children);
+
+function createShallowNodeMock() {
+  let isFirst = true;
+  return function createNodeMock(element) {
+    if (isFirst) {
+      isFirst = false;
+      return element.type;
+    }
+    return ShallowNodeMockComponent;
+  };
+}
+
+function wrapElementWithContextProvider(element, context) {
+  const childContextTypes = Object.keys(context).reduce(
+    (reduced, key) => {
+      reduced[key] = emptyFunction;
+      return reduced;
+    },
+    {},
+  );
+
+  class ShallowRendererWrapper extends React.Component {
+    static childContextTypes = childContextTypes;
+    getChildContext() {
+      return context;
+    }
+    render() {
+      return this.props.children;
+    }
+  }
+
+  return React.createElement(ShallowRendererWrapper, null, element);
+}
+
+class ReactShallowRenderer {
+  getMountedInstance() {
+    return this._renderer ? this._renderer.getInstance() : null;
+  }
+
+  getRenderOutput() {
+    if (this._renderer) {
+      const tree = this._renderer.toTree();
+      if (tree && tree.rendered) {
+        // If we created a context-wrapper then skip over it.
+        const element = tree.type.childContextTypes
```

Comment:
This seems fragile. What if it isn't *our* context but the root element just happens to be a context provider?

To be honest I'm not a fan of second undocumented `context` argument. Our other APIs don't take it. I think it's remaining from the time context was owner-based and we might as well deprecate it and tell people to use their own context providers explicitly. But it's probably annoying to do in this PR.

Let's just be stricter with checking here.